tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -199,41 +199,100 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
   }
 
   /** Desugars parameters and body to a JS function.
    */
-  def desugarToFunction(
-      params: List[ParamDef],
-      body: Tree, isStat: Boolean)(
-      implicit globalKnowledge: GlobalKnowledge, pos: Position): js.Function = {
+  def desugarToFunction(params: List[ParamDef], body: Tree, isStat: Boolean)(
+      implicit globalKnowledge: GlobalKnowledge,
+      pos: Position): WithGlobals[js.Function] = {
     new JSDesugar().desugarToFunction(params, body, isStat, Env.empty)
   }
 
   private class JSDesugar()(implicit globalKnowledge: GlobalKnowledge) {
 
-    // Synthetic variables
+    // Name management
+
+    /** Whether we are running in the "optimistic naming" run.
+     *
+     *  In theory, `JSDesugar` works in two passes: the optimistic run,
+     *  followed by the pessimistic run should the first one fail.
+     *
+     *  The optimistic run assumes that there is no clash between a local
+     *  variable name and a global variable name (i.e., a `JSGlobalRef`). This
+     *  allows it to straightforwardly translate IR identifiers to JS
+     *  identifiers. While it does that, it records all the local variable and
+     *  global variable names that are used in the method.
+     *
+     *  At the end of the translation, we check whether there was a clash by
+     *  testing if the two sets intersect. If they do not, we are lucky, and
+     *  can completely by-pass the pessimistic run. If there is a clash, then
+     *  we need to restart everything in pessimistic mode.
+     *
+     *  In the pessimistic run, we use the set of global variable names that
+     *  was collected during the optimistic run to *prevent* clashes from
+     *  happening. This requires that we maintain a map of IR identifiers to
+     *  allocated JS identifiers, as some IR identifiers need to be renamed.
+     */
+    private var isOptimisticNamingRun: Boolean = true
+
+    private val globalVarNames = mutable.Set.empty[String]
+    private val localVarNames = mutable.Set.empty[String]
+
+    private lazy val localVarAllocs = mutable.Map.empty[String, String]
+
+    private def referenceGlobalName(name: String): Unit =
+      globalVarNames += name
+
+    private def extractWithGlobals[A](withGlobals: WithGlobals[A]): A = {
+      for (globalRef <- withGlobals.globalVarNames)
+        referenceGlobalName(globalRef)
+      withGlobals.value
+    }
+
+    private def transformLocalName(name: String): String = {
+      if (isOptimisticNamingRun) {
+        localVarNames += name
+        name
+      } else {
+        def slowPath(): String = {
+          localVarAllocs.getOrElseUpdate(name, {
+            var suffix = 0
+            var result = name
+            while (globalVarNames.contains(result) ||
+                localVarNames.contains(result)) {
+              suffix += 1
+              result = name + "$" + suffix
+            }
+            localVarNames += result
+            result
```

Comment:
No in fact it's much more stupid than that. It's obviously not quadratic because this piece of code is guarded by `localVarAllocs.getOrElseUpdate(name, { ... })`. So clearly the loop there is only executed once per input `name`, and not _n_ times as was the case in the optimizer.

The core reason is still what I said: in the optimizer, we create *fresh* names. We invent many *different* fresh names based on the same `base`. Here we only need to allocate one name per input `name`.

It also becomes clear if you consider that the *fix* to the optimizer was to basically add such a `getOrElseUpdate(base, ...)`. There's nothing *more* that we could do here.